src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -1846,27 +1846,33 @@ private static object DeserializeMessageHeaderHelper(BinaryTokenStreamReader str
             throw new SerializationException(String.Format("Unexpected token {0} parsing message headers", token));
         }
 
-        private static bool TryFindAndRegisterExternalSerializer(Type t, out IExternalSerializer serializer)
+        private static bool TryLookupExternalSerializer(Type t, out IExternalSerializer serializer)
         {
-            if (externalSerializers.Count == 0 || typesWithNoExternalSerializers.ContainsKey(t))
+            // essentially a no-op if there are no external serializers registered
+            if (externalSerializers.Count == 0)
             {
                 serializer = null;
                 return false;
             }
 
-            serializer = externalSerializers.FirstOrDefault(s => s.IsSupportedType(t));
-            if (serializer == null)
+            // the associated serializer will be null if there are no external serializers that handle this type
+            if (typeToExternalSerializerDictionary.TryGetValue(t, out serializer))
             {
-                typesWithNoExternalSerializers.TryAdd(t, 0);
-                return false;
+                return serializer != null;
             }
 
-            Register(
-                t,
-                serializer.DeepCopy,
-                serializer.Serialize,
-                serializer.Deserialize);
-            return true;
+            serializer = externalSerializers.FirstOrDefault(s => s.IsSupportedType(t));
+
+            // add the serializer to the dictionary, even if it's null to signify that we already performed
+            // the search and found none
+            typeToExternalSerializerDictionary.TryAdd(t, serializer);
```

Comment:
if `TryAdd` returns false we shouldn't call `Register`, even if `serializer` is non-null